### ISSUE TITLE
ttnn.hypot returns inf above |x| ~= 1.8e19 and 0 below ~1.1e-19, where the bfloat16 golden is finite

### ISSUE THREAD

**Issue by bvdyvncqr5-commits** [tenstorrent/tt-metal#51861]
# ttnn.hypot returns inf above |x| ~= 1.8e19 and 0 below ~1.1e-19, where the bfloat16 golden is finite

`ttnn.hypot(1e25, 0)` returns `inf`. For the bfloat16 input below, the registered golden is the
finite value `9.9736380118206907e24`.

`hypot(x, 0)` is `|x|` by definition. There is no rounding question, no tolerance to argue about, and
no accuracy trade to weigh — and the registered golden agrees:

```python
torch.hypot(torch.tensor([1e25], dtype=torch.bfloat16),
            torch.tensor([0.0],  dtype=torch.bfloat16))   # 9.9736380118206907e+24, finite
```

The op is composed so that it computes `sqrt(a**2 + b**2)` literally
(`binary_ng_utils.cpp:391-396`):

```c
case BinaryOpType::HYPOT:
    process_lhs = unary::UnaryOpType::SQUARE;
    process_rhs = unary::UnaryOpType::SQUARE;
    binary_op = EnumT::ADD;
    postprocess = unary::UnaryOpType::SQRT;
    break;
```

`a**2` overflows or enters Tenstorrent's DAZ/FTZ region long before `hypot(a, b)` does. Avoiding
those intermediates is why `hypot` exists as a named function rather than an expression callers
write out.

| input (bfloat16) | `ttnn.hypot` | registered golden | golden finite? |
|---|---|---|---|
| `(1e25, 0)` | `inf` | `9.9736380118206907e24` | yes |
| `(3e38, 0)` | `inf` | `3.0040552704739099e38` | yes |
| `(1e25, 1e25)` | `inf` | `1.4129320516745978e25` | yes |
| `(1e30, 1e30)` | `inf` | `1.416203404942475e30` | yes |
| `(1e20, 1e20)` | `inf` | `1.4123288431433875e20` | yes |
| `(1.3835e19, 1.3835e19)` | `inf` | `1.95996655783164e19` | yes — **and both operands are below `2^64`** |
| `(1.3835e19, 0)` | `1.3835058055282164e19` | `1.3835058055282164e19` | — (same operand, correct when paired with `0`) |
| `(1e-20, 1e-20)` | `0.0` | `1.4187801866509531e-20` | yes |
| `(3, 4)` | `5` | `5` | — |

Taking one operand at a time, the observed boundaries are exactly where squaring overflows or enters
the accelerator's FTZ region. Sweeping every binade with five mantissas each (1.00, 1.25, 1.50, 1.75, 1.99 — 20 of 25 sample points
are not powers of two) puts the boundaries at a single bfloat16 value:

| side | last value that works | first value that fails |
|---|---|---|
| overflow | `bfloat16(1.99 x 2^63)` = `1.9921875 x 2^63` = `1.8374686479671624e19` | **`2^64`** = `1.84467e19` (`inf`; golden `1.844674e19`) |
| underflow | `2^-63` = `1.08420e-19` | `bfloat16(1.99 x 2^-64)` = `1.9921875 x 2^-64` = `1.0799670077492329e-19` (`0`; golden `1.079967e-19`) |

For a single nonzero normal bfloat16 operand, the failing set is `|x| >= 2^64` or `0 < |x| < 2^-63`.

Two operands fail sooner than one, because `a**2 + b**2` reaches the overflow point before `a**2`
does. Sweeping equal-magnitude pairs the same way puts the onset at or below `1.5 x 2^63` — below
the single-operand threshold — and the contrast is direct: `hypot(1.5 x 2^63, 1.5 x 2^63)` returns
`inf` while the same operand paired with `0` returns the correct value. Three sampled pairs sit in
that gap, and the sweep brackets the true onset in `(1.25 x 2^63, 1.5 x 2^63]`. Identical results on
Wormhole and Blackhole; ordinary magnitudes (`3`, `5`, `100`, each paired with itself and with `0`)
pass in the same run as a control.

Those single-operand boundaries match the overflow and normal-range/FTZ boundaries predicted by
squaring. On the low side,
`sqrt(1.1754943508222875e-38)` is exactly `2^-63`: its square is the smallest normal bfloat16,
while the next sampled value squares to a bfloat16 subnormal. Tenstorrent's own tests state that
hardware DAZ/FTZ flushes a subnormal intermediate to zero
(`tests/ttnn/unit_tests/operations/eltwise/test_activation.py:135-141`). This does not challenge
that documented behavior: the composite unnecessarily materialises an FTZ intermediate although
the final `hypot` golden is a normal finite bfloat16 value. On the overflow side, the algebraic threshold,
`sqrt(3.3895313892515355e38) = 1.84107e19`, falls *between* the last working value and the first
failing one, because bfloat16 has few representable values at that magnitude. **`2^64` is the
measured boundary; `1.841e19` is only the algebraic one.**

**Reference values are the registered golden** (`ttnn/ttnn/operations/binary.py`,
`attach_golden_function`) — that is `torch.hypot`, evaluated with **both arguments at the operator's
own dtype (bfloat16)**, not on upcast tensors. The golden is finite at every input above; it does
not share the overflow.

The defining identity holds on the golden across the whole range, which is what makes the failure
unambiguous rather than a tolerance question:

| x | `torch.hypot(x, 0)` | `\|x\|` |
|---|---|---|
| `1e-25` | `1.001664023013432e-25` | equal |
| `1.0` | `1` | equal |
| `1e19` | `1.0016005571271983e19` | equal |
| `1e25` | `9.9736380118206907e24` | equal |
| `3e38` | `3.0040552704739099e38` | equal |

Five of the nine magnitudes sampled are shown; equality is exact at every one of the nine.

## This is a hazard the repository already names and avoids elsewhere

`ckernel_sfpu_trigonometry.h` states the same number and the same cause, identically on all three
architectures — `tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_trigonometry.h:1057-1058`,
`tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_trigonometry.h:1021-1022`,
`tt_metal/hw/ckernels/quasar/metal/llk_api/llk_sfpu/ckernel_sfpu_trigonometry.h:1006-1007`:

```
// acosh(x) == ln(2x) to <1 ulp; using log1p(2x - 1) also dodges the x^2 overflow
// that makes the classic form return +inf for x >= ~1.84e19.
```

`acosh` was restructured to avoid squaring. `hypot` — the one function whose entire
purpose is to avoid it — was not.

## Reproduction

```python
import torch, ttnn
a = torch.tensor([[1e25, 1e20, 1.3835e19, 1e-20]], dtype=torch.bfloat16)
b = torch.tensor([[0.0,  1e20, 1.3835e19, 1e-20]], dtype=torch.bfloat16)
device = ttnn.open_device(device_id=0)
mk = lambda t: ttnn.from_torch(t, dtype=ttnn.bfloat16, device=device, layout=ttnn.TILE_LAYOUT)
print(ttnn.to_torch(ttnn.hypot(mk(a), mk(b))))    # inf, inf, inf, 0.0
print(ttnn.get_golden_function(ttnn.hypot)(a, b)) # 9.9736e+24, 1.4123e+20, 1.9600e+19, 1.4188e-20
# the third column is the pair case: both operands are below 2^64, and only the pair overflows
```

Measured on ttsim, Wormhole and Blackhole, bfloat16 — identical on both. I do not have a device;
these are simulator runs, not silicon measurements.

## Relationship to #29406

[PR #29406](https://github.com/tenstorrent/tt-metal/pull/29406) ("#27654: Migrate Hypot as device
op", merged 2025-10-10) removed the host-level composite and introduced the
`SQUARE / SQUARE / ADD / SQRT` block quoted above, reporting a 67% kernel-duration improvement. It
also added two hypot tests.

So this is not a suggestion to make hypot a device op — that landed nine months ago. It is that the
decomposition chosen in that migration squares both operands, and that its own acceptance tests stop
well short of where squaring overflows or enters the normal-range/FTZ boundary:

| test (both added by #29406) | largest magnitude | assertion |
|---|---|---|
| `tests/ttnn/unit_tests/operations/eltwise/test_binaryng_fp32.py:164` (`test_hypot_multi_dtype`) | `999.999` (`test_binaryng_fp32.py:170`, a fixed literal list) | `test_binaryng_fp32.py:180` `assert_with_ulp(ulp_threshold=1)`; `test_binaryng_fp32.py:178` `assert_allclose(rtol=1e-5)` |
| `tests/ttnn/unit_tests/operations/eltwise/test_binaryng_ND.py:71` (in the op list), `test_binaryng_ND.py:101-102` | 50 | `test_binaryng_ND.py:102` `assert_with_ulp(ulp_threshold=2)` |

The coverage is stricter than PCC — ULP-1 and ULP-2 — but the largest magnitude exercised is `1e3`,
against the measured onset of `2^64` (`1.84467e19`): sixteen orders of magnitude away. The band is
outside test coverage rather than inside it with a tolerance that is too loose. On the small side,
the ND test constructs bfloat16 values with `torch.rand(..., dtype=torch.bfloat16) * 100 - 50`
(`test_binaryng_ND.py:77,82`); its smallest nonzero magnitude is `0.25`, so it cannot emit this
band at all.

## Which dtypes take this path

`binary_ng_device_operation.cpp:33-37` routes `HYPOT` to the SFPU path only when **both** inputs are
float32. It shares that rule with four other composed ops:

```c
case LOGADDEXP:
case LOGADDEXP2:
case LDEXP:
case BIAS_GELU:
case HYPOT: return (a == FLOAT32 && b == FLOAT32);
```

So bfloat16 takes the FPU composite quoted above and float32 takes a different path. Everything in
this report is measured on bfloat16. `test_hypot_multi_dtype` parametrises both dtypes
(`test_binaryng_fp32.py:163-164`), so a fix should say which one it targets; I have only measured
bfloat16.

The same `SQUARE / SQUARE / ADD / SQRT` block also appears in the quasar tree
(`ttnn/cpp/ttnn/operations/experimental/quasar/binary_ng/device/binary_ng_utils.cpp:391-396`).

## Proposed fix

Use the standard technique: scale both operands by a power of two before squaring, and scale the
result back. Powers of two are exact in binary floating point, so nothing is lost in the scaling
itself — it only moves where the intermediate lands. The guard has to sit below the point where the
sum of *two* squares can overflow, not merely where one can.

Two consequences worth stating up front rather than having you find them:

- Whether inputs that work today stay bit-identical depends on the placement. Pure power-of-two
  scaling inside the existing composite changes nothing for them; a dedicated path with different
  intermediate precision can move currently-passing results — no less accurate, but not identical.
  That is one more reason the placement is a maintainer call rather than a detail.
- Placement is a real decision, not an oversight in this report. The bfloat16 op takes the FPU
  composite path shown above, so the change belongs either in that decomposition or in a dedicated
  path, and the two have different cost profiles. I have sanity-checked the scaled arithmetic on
  ttsim, but through a different dispatch class — a single SFPU kernel with fp32 register
  intermediates — not through the bfloat16 FPU-composite path this report measures, and I have no
  way to exercise the host-side wiring in simulation at all. So the arithmetic is checked; the
  placement is not, and the choice of placement determines what I can validate before a PR.

## Verification scope

The failing bfloat16 operation itself was reproduced end to end on ttsim for both Wormhole and
Blackhole, against the operator-dtype `torch.hypot` golden shown above. These are simulator runs,
not silicon measurements. The report makes no claim about a final integration, its throughput, or
the float32 and special-value paths.

## Success criteria

- [ ] `hypot(x, 0) == |x|` exactly, for every normal bfloat16 `x`, on both architectures — this
      family needs no tolerance argument at all
- [ ] both measured single-operand bands — `|x| >= 2^64` and `0 < |x| < 2^-63` — return a finite
      result matching the golden within the ULP thresholds this repository already applies to
      `hypot` (1 and 2, from #29406)
- [ ] the same for pairs where **each** operand is below the single-operand onset but the sum of
      squares still overflows — measured onset at or below `1.5 x 2^63`, so a guard placed at `2^64`
      would satisfy the previous item and still return `inf` here
- [ ] ordinary inputs keep passing the two tests added by #29406, at their present thresholds
- [ ] bfloat16 test coverage extended past `1e3`, which is where the current sweeps stop, so the
      band cannot silently reappear
- [ ] the same checks run on Blackhole in its own right, not inferred from Wormhole

Scope limits are listed below.

## What this does not settle

- The float32 path. It takes a different SFPU path and was not investigated.
- `hypot(inf, x)` and `hypot(nan, x)`. Not measured; a future implementation would need to specify
  its handling of them.
- Instruction count and throughput. I have not measured either; the final integration path has not
  yet been selected.
- Whether this belongs as a follow-up to #29406 or as a separate change.

## Next step

The failing bands, the formulation and the test gap above are worked out. The placement is not, and
it is the piece that needs a maintainer's call — it also decides how much I can validate before
opening anything. I can prepare a PR with regression coverage for both bands and for ordinary
inputs, and would welcome a steer on which integration point you prefer.

**Comment by bvdyvncqr5-commits** [comment]
Filling in the bug report template fields. The original write-up was long and had no runnable repro — that was my mistake.

### Component / Area

kernels (eltwise binary)

### Issue Type (optional)

Bad Outputs

### Observed

`ttnn.hypot(x, x)` returns `inf` for `x = 1.3835058055282164e19`.

### Expected

`1.95996655783164e+19`

Both operands are `1.38e19` and the answer is finite and well inside bfloat16 range. The same operand paired with `0` comes back correct, so what overflows is the intermediate `x*x`, not the input and not the result.

### 1. Steps (exact commands)

```python
import torch, ttnn

device = ttnn.open_device(device_id=0)
t = torch.full((32, 32), 1.3835058055282164e19, dtype=torch.bfloat16)  # 1.5 * 2**63
z = torch.zeros((32, 32), dtype=torch.bfloat16)
a = ttnn.from_torch(t, layout=ttnn.TILE_LAYOUT, device=device)
b = ttnn.from_torch(z, layout=ttnn.TILE_LAYOUT, device=device)

print(ttnn.to_torch(ttnn.hypot(a, a))[0, 0].item())   # inf
print(ttnn.to_torch(ttnn.hypot(a, b))[0, 0].item())   # 1.3835058055282164e+19
print(torch.hypot(t, t)[0, 0].item())                 # 1.95996655783164e+19
```

### 2. Input data / link or description

None beyond the two constants above.

### 3. Frequency

Always, every element of the tile.

### 1. Software Versions

Ubuntu 22.04.5, ttnn 0.75.0rc5, ttsim v1.9.6.

### 2. Hardware Details

None — I do not have Tenstorrent hardware. Run on ttsim against both the Wormhole and Blackhole targets; the three printed values are identical on both.

### Is this a regression?

Unknown.

### How it was detected

A differential sweep running ttnn eltwise ops against their registered torch goldens over corner values of the bfloat16 grid, with `abs` / `neg` / `identity` / `relu` as a control group.

**Comment by EazyHood** [comment]
The float32 path you listed under "what this does not settle" fails the same way, at the same two boundaries — I think the dispatch reading in the report is slightly off, and it makes the fix simpler rather than harder.

`is_binary_sfpu_op` does route `HYPOT` to the SFPU only when both inputs are float32, so the *engine* differs. But the decomposition does not. `OpConfig::OpConfig` is a template instantiated once per engine (`binary_ng_utils.cpp:698-699`), and cases that need per-engine handling ask for it explicitly — `DIV` at `:159`, `RSUB` at `:172`, 26 such branches in that switch. `HYPOT` at `:391-396` has none:

```cpp
case BinaryOpType::HYPOT:
    process_lhs = unary::UnaryOpType::SQUARE;
    process_rhs = unary::UnaryOpType::SQUARE;
    binary_op   = EnumT::ADD;
    postprocess = unary::UnaryOpType::SQRT;
    break;
```

So float32 squares too, and the boundaries land on the same numbers as your bfloat16 measurements — `sqrt(FLT_MAX)` is `1.8446744e19` = `2^64` and `sqrt(FLT_MIN)` is `1.0842022e-19` = `2^-63`. That is not a coincidence: bfloat16 and float32 share the same 8-bit exponent range, and whether a square overflows depends only on the exponent. Your two thresholds are properties of the exponent range, not of bfloat16.

Reimplementing the composite as written in float32 on x86, against a range-safe hypot, over five mantissas per binade from `1e-30` to `1e30` with and without a second operand:

```
610 cases -> 416 correct, 194 broken (31.8%)
hypot(1e25, 0)      inf   (exact answer 9.999999e24)
hypot(3e38, 0)      inf   (exact answer 3.000000e38)
hypot(1e25, 1e25)   inf   (exact answer 1.414213e25)
hypot(1e-25, 0)     0     (exact answer 1.000000e-25)
hypot(3, 4)         5     control, unchanged
```

Control: the range-safe formulation returns a finite non-zero result at every one of those magnitudes, so the zeros and infinities are the composite's, not the harness's.

I have no device either; this is the composite reimplemented on x86, so treat it as a reading of the decomposition rather than a silicon measurement.

Where I think this helps: the placement question you flagged as needing a maintainer's call gets easier if there is only one decomposition rather than two. Fixing `OpConfig`'s `HYPOT` case covers both engines at once, and the `is_sfpu_op()` branches already in that switch are the precedent for splitting it later if the two engines want different treatment.
